tensorboard/data/server/run.rs

```diff
@@ -0,0 +1,363 @@
+    files: BTreeMap<PathBuf, EventFile<BufReader<File>>>,
+
+    /// Reservoir-sampled data and metadata for each time series.
+    time_series: HashMap<Tag, TimeSeries>,
+}
+
+#[derive(Debug)]
+enum EventFile<R> {
+    /// An event file that may still have more valid data.
+    Active(EventFileReader<R>),
+    /// An event file that can no longer be read.
+    ///
+    /// This can be because of a non-recoverable read error (e.g., a bad length checksum), due to
+    /// the last-read record being very old (note: not yet implemented), or due to the file being
+    /// deleted.
+    Dead,
+}
+
+#[derive(Debug)]
+struct TimeSeries {
+    data_class: pb::DataClass,
+    metadata: Box<pb::SummaryMetadata>,
+    rsv: StageReservoir<StageValue>,
+}
+
+impl TimeSeries {
+    fn new(metadata: Box<pb::SummaryMetadata>) -> Self {
+        let data_class =
+            pb::DataClass::from_i32(metadata.data_class).unwrap_or(pb::DataClass::Unknown);
+        let capacity = match data_class {
+            pb::DataClass::Scalar => 1000,
+            pb::DataClass::Tensor => 100,
+            pb::DataClass::BlobSequence => 10,
+            _ => 0,
+        };
+        Self {
+            data_class,
+            metadata,
+            rsv: StageReservoir::new(capacity),
+        }
+    }
+}
+
+/// A value staged in the reservoir.
+///
+/// This is kept as close as possible to the on-disk event representation, since every record in
+/// the stream is converted into this format.
+#[derive(Debug)]
+struct StageValue {
+    wall_time: WallTime,
+    payload: StagePayload,
+}
+
+#[derive(Debug)]
+enum StagePayload {
+    #[allow(dead_code)]
+    GraphDef(Vec<u8>),
+    Summary(Box<pb::summary::value::Value>),
+}
+
+impl RunLoader {
+    pub fn new() -> Self {
+        Self {
+            start_time: None,
+            files: BTreeMap::new(),
+            time_series: HashMap::new(),
+        }
+    }
+
+    /// Loads new data given the current set of event files.
+    ///
+    /// The provided filenames should correspond to the entire set of event files currently part of
+    /// this run.
+    pub fn reload(&mut self, filenames: Vec<PathBuf>) {
+        self.update_file_set(filenames);
+        self.reload_files();
+    }
+
+    /// Updates the active key set of `self.files` to match the given filenames.
+    ///
+    /// After this function returns, `self.files` may still have keys not in `filenames`, but they
+    /// will all map to [`EventFile::Dead`].
+    fn update_file_set(&mut self, filenames: Vec<PathBuf>) {
+        // Remove any discarded files.
+        let new_file_set: HashSet<&PathBuf> = filenames.iter().collect();
+        for (k, v) in self.files.iter_mut() {
+            if !new_file_set.contains(k) {
+                *v = EventFile::Dead;
+            }
+        }
+
+        // Open readers for any new files.
+        for filename in filenames {
+            use std::collections::btree_map;
```

Comment:
Yeah, this is fine!

First, note that Rust imports are always pure. Modules do not have side
effects. So you can freely reorder imports (`rustfmt` sorts them) or
declare them at any scope that you like without worrying about bugs.

So, then, it’s all about scope. The general Rust convention is to import
most types and traits that you use at top level, and modules if you
want, too. Importing bare functions is rarer, I think, but not unheard
of. I like to be able to see `data_compat::initial_metadata(...)` to
clarify that it’s not just a helper function in this module.

So I could have imported `btree_map::Entry` at top level. But I think
that that would be a wee bit confusing, because if you see `Entry` in
some arbitrary place in the code, it’s not obvious what that refers to
(as opposed to, e.g., `HashMap`).

I also could have imported `btree_map` at top level, and that would also
have been reasonable imho. I don’t really have strong opinions here.

Some examples of my conventions:

```rust
// OK: types and traits that are clear from context
use std::collections::{HashMap, HashSet};
use std::io::{Read, Write};

// OK: modules, aliases
use std::io;
use std::path;
use crate::proto::tensorboard as pb;

// Avoid: types not clear from context
// use std::io::Error;  // a fn returns "Error"---what kind of error?

// Weak-avoid: standalone functions
// use std::iter::once;  // seeing "iter::once" would be clearer
```